cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/DynamicPartitioner.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © 2014-2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.dataset.lib;
+
+import co.cask.cdap.api.mapreduce.MapReduceTaskContext;
+
+/**
+ * Responsible for dynamically determining a @{link PartitionKey}.
+ * For each K, V pair, the getPartitionKey(K, V) method is called to determine a PartitionKey.
+ *
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public abstract class DynamicPartitioner<K, V> {
+
+
+  /**
+   *  Initializes a DynamicPartitioner.
+   *  <p>
+   *    This method will be called only once per {@link DynamicPartitioner} instance. It is the first method call
+   *    on that instance.
+   *  </p>
+   *  @param mapReduceTaskContext the mapReduceTaskContext for the task that this DynamicPartitioner is running in.
+   *  Note that the hadoop context is not available on this MapReduceTaskContext.
+   */
+  public void initialize(MapReduceTaskContext<K, V> mapReduceTaskContext) {
+    // do nothing by default
+  }
+
+  /**
+   *  Destroys a DynamicPartitioner.
+   *  <p>
+   *    This method will be called only once per {@link DynamicPartitioner} instance. It is the last method call
+   *    on that instance.
+   *  </p>
+   */
+  public void destroy() {
```

Comment:
Hmmm, we have a ``Destroyable`` interface in ``cdap-etl-api`` that defines a ``destroy()`` method. Should we move it to ``cdap-api`` and re-use instead of re-defining this method here? Maybe not this PR, but a jira?